tests/cpp/test_frontend.cc

```diff
@@ -5,16 +5,498 @@
+  auto input_stream = rapidjson::MemoryStream(json_str.c_str(),
+                                              json_str.size());
+  std::shared_ptr<MockDelegator> delegator =
+    std::make_shared<MockDelegator>();
+
+  std::string output;
+  details::ObjectiveHandler wrapped_handler {delegator, output};
+  MockObjectStarter handler {delegator, wrapped_handler};
+
+  rapidjson::Reader reader;
+  auto result = reader.Parse(input_stream, handler);
+
+  ASSERT_TRUE(result);
+
+  ASSERT_EQ(output, "multi:softmax");
+}
+
+/******************************************************************************
+ * LearnerParamHandler
+ * ***************************************************************************/
+TEST(LearnerParamHandlerSuite, LearnerParamHandler) {
+  std::string json_str =
+    "{\"base_score\": \"0.5\", \"num_class\": \"5\", \"num_feature\": \"12\"}";
+  auto input_stream = rapidjson::MemoryStream(json_str.c_str(),
+                                              json_str.size());
+  std::shared_ptr<MockDelegator> delegator =
+    std::make_shared<MockDelegator>();
+
+  Model output;
+  details::LearnerParamHandler wrapped_handler {delegator, output};
+  MockObjectStarter handler {delegator, wrapped_handler};
+
+  rapidjson::Reader reader;
+  auto result = reader.Parse(input_stream, handler);
+
+  ASSERT_TRUE(result);
+
+  ASSERT_FLOAT_EQ(output.param.global_bias, 0.5);
+  ASSERT_EQ(output.num_output_group, 5);
+  ASSERT_EQ(output.num_feature, 12);
+}
+
+/******************************************************************************
+ * XGBoostModelHandler
+ * ***************************************************************************/
+TEST(XGBoostModelHandlerSuite, XGBoostModelHandler) {
+  std::string json_str = "{\"version\": []}";
+  auto input_stream = rapidjson::MemoryStream(json_str.c_str(),
+                                              json_str.size());
+  std::shared_ptr<MockDelegator> delegator =
+    std::make_shared<MockDelegator>();
+
+  Model output;
+  details::LearnerParamHandler wrapped_handler {delegator, output};
+  MockObjectStarter handler {delegator, wrapped_handler};
+
+  rapidjson::Reader reader;
+  auto result = reader.Parse(input_stream, handler);
+}
+
+/******************************************************************************
+ * RootHandler
+ * ***************************************************************************/
+TEST(RootHandlerSuite, RootHandler) {
   std::string json_str = "{}";
+  auto input_stream = rapidjson::MemoryStream(json_str.c_str(),
+                                              json_str.size());
+  std::shared_ptr<MockDelegator> delegator =
+    std::make_shared<MockDelegator>();
+
+  Model output;
+  details::LearnerParamHandler wrapped_handler {delegator, output};
+  MockObjectStarter handler {delegator, wrapped_handler};
 
-  Model model;
-  frontend::LoadXGBoostJSONModel(json_str, &model);
+  rapidjson::Reader reader;
+  auto result = reader.Parse(input_stream, handler);
```

Comment:
This test does not have any assertion. Is it sufficient that the test runs without crashing?

Reply:
Same story as above. This one is even higher up the JSON tree, so it should be fully exercised by the integration test and has no non-trivial tests that can be performed in isolation.